lib/PreviewProvider.js

```diff
@@ -0,0 +1,239 @@
+    */
+  cleanUpCache: Task.async(function*() {
+    let currentTime = Date.now();
+    Object.keys(ss.storage.embedlyData).forEach(item => {
+      if (currentTime - ss.storage.embedlyData[item].accessTime > CLEAR_CACHE_TIMEOUT) {
+        delete ss.storage.embedlyData[item];
+      }
+    });
+    this._cacheTimeout();
+  }),
+
+  /**
+    * Set up the timer for the cache to be cleaned every 24 hours
+    */
+  _cacheTimeout() {
+    this._timeoutID = setTimeout(() => {
+      this.cleanUpCache();
+    }, CLEAR_CACHE_TIMEOUT);
+  },
+
+  /**
+    * Filter out sites that do not have acceptable protocols and hosts
+    */
+  _URLFilter(definition) {
+    return function(item) {
+      let result = true;
+      definition.forEach((test) => {
+        if (!test(item)) {
+          result = false;
+        }
+      });
+      return result;
+    };
+  },
+
+  /**
+    * Santize the URL to remove any unwanted or sensitive information about the link
+    */
+  _sanitizeURL(site) {
+    if (!site) {
+      return "";
+    }
+
+    let queryParams = new Map();
+    let safeQueryParam = "";
+    let url = new URL(site);
+
+    if (url.search.length === 0) {
+      return url.toString();
+    }
+    // extract and parse the query parameters, if any
+    url.search.slice(1, url.search.length);
+    let paramArray = url.search.slice(1, url.search.length).split("&");
+    paramArray.forEach(item => {
+      let tmp = item.split("=");
+      queryParams.set(tmp[0], tmp[1]);
+    });
+    // remove the unwanted parameters and update the query string
+    REMOVE_KEYS.forEach(key => queryParams.delete(key));
+    queryParams.forEach((val, key) => {
+      if (val.length !== 0 && key.length !== 0) {
+        safeQueryParam = safeQueryParam + key + "=" + val + "&";
+      }
+    });
+    url.search = safeQueryParam.slice(0, safeQueryParam.length - 1);
+    url.hash = "";
+
+    // remove extra slashes in pathname and construct back a safe pathname
+    if (url.pathname) {
+      let pathItems = url.pathname.split("/");
+      let safePathItems = pathItems.filter(item => item.length !== 0);
+      let safePath = "/";
+      url.pathname = safePath + safePathItems.join("/");
+    }
+
+    // if the url doesn't have a protocol, use https
+    if (!url.protocol) {
+      url.protocol = "https:";
+    }
+
+    return url.toString();
+  },
+
+  /**
+    * Create a key based on a URL in order to dedupe sites
+    */
+  _createKey(url) {
+    url = new URL(url);
+    let key = url.host.replace(/www\.?/, "");
+    key = key + url.pathname;
+    return key;
+  },
+
+  /**
+    * Canonicalize urls by deduping them, then sanitizing them
+    */
+  _dedupeURLs(sites) {
+    let dedupedSites = new Map();
+    sites.forEach(site => {
+      let urlkey = this._createKey(site.url);
+      if (!dedupedSites.has(urlkey)) {
+        dedupedSites.set(urlkey, site);
+      } else {
+        // if two sites come in with 'http' and 'https', pick 'https'
```

Comment:
I'm worried about two things here

1) separating and potentially duplicating logic between the client and server with respect to how we decide to group or cluster URLs around protocol or domain or what have you.  I'm in favour of the client filtering out potentially sensitive data from URLs such as query strings which might contain private user IDs, but removing protocol can actually introduce issues and 

2) I've already witnessed situations where embedly's responses will vary solely based on protocol, so I think it might be advisable to actually preserve both URLs rather than deduping them.